keps/sig-security/2763-ambient-capabilities/README.md

```diff
@@ -0,0 +1,920 @@
+  AddCapabilities []string `protobuf:"bytes,1,rep,name=add_capabilities,json=addCapabilities,proto3" json:"add_capabilities,omitempty"`
+  // List of capabilities to drop.
+  DropCapabilities     []string `protobuf:"bytes,2,rep,name=drop_capabilities,json=dropCapabilities,proto3" json:"drop_capabilities,omitempty"`
+  XXX_NoUnkeyedLiteral struct{} `json:"-"`
+  XXX_sizecache        int32    `json:"-"`
++ // List of ambient capabilities to add.
++ AddAmbientCapabilities []string `protobuf:"bytes,1,rep,name=add_ambient_capabilities,json=addAmbientCapabilities,proto3" json:"add_ambient_capabilities,omitempty"`
+}
+```
+
+Adding ambientCapabilities to the SecurityContext will directly update this field, and these changes would live in the [convertToRuntimeCapabilities](https://github.com/kubernetes/kubernetes/blob/ea0764452222146c47ec826977f49d7001b0ea8c/pkg/kubelet/kuberuntime/security_context.go#L121:6) function.
+
+
+### Changes to containerd/containerd (https://github.com/containerd/containerd)
+
+`containerd` clears all ambient capabilities on container creation. See [this](https://github.com/containerd/containerd/blob/055c801ededcb7a5e82f47bdeed555cdf6c64bd8/pkg/cri/server/container_create_linux.go#L233) link for details. 
+
+```go
+// Clear all ambient capabilities. The implication of non-root + caps
+// is not clearly defined in Kubernetes.
+// See https://github.com/kubernetes/kubernetes/issues/56374
+// Keep docker's behavior for now.
+specOpts = append(specOpts,
+	customopts.WithoutAmbientCaps,
+	customopts.WithSelinuxLabels(processLabel, mountLabel),
+)
+```
+
+We would need to update code at that link to add the ambient capabilities by calling the [WithAmbientCapabilities](https://github.com/containerd/containerd/blob/ab963e1cc16a845567a0e3e971775c29c701fcf8/oci/spec_opts.go#L858) function instead of the [WithoutAmbientCapabilities](https://github.com/containerd/containerd/blob/04f73e3f8a097d95111f8419fa136d196b3a8725/pkg/cri/opts/spec_linux.go#L354) function. 
+
+### Order of changes
+We propose that the changes be made in the following order:
+
+1. Changes described in the [Changes to runtime API (<a href="https://pkg.go.dev/k8s.io/cri-api/pkg/apis/runtime/v1alpha2">https://pkg.go.dev/k8s.io/cri-api/pkg/apis/runtime/v1alpha2</a>)](#changes-to-runtime-api-httpspkggodevk8siocri-apipkgapisruntimev1alpha2) section are made first as both containerd and kubernetes rely on these.
+2. Changes described in the [Changes to containerd/containerd (<a href="https://github.com/containerd/containerd">https://github.com/containerd/containerd</a>)](#changes-to-containerdcontainerd-httpsgithubcomcontainerdcontainerd) are made next.
+3. Once a version of `containerd` that supports ambient capabilities is available we make the changes described in the [Changes to kubernetes API (<a href="https://pkg.go.dev/k8s.io/api/core/v1">https://pkg.go.dev/k8s.io/api/core/v1</a>)](#changes-to-kubernetes-api-httpspkggodevk8sioapicorev1) section.
+
+
```

Comment:
In a real world scenario, different teams are responsible for node OS ( by corollary, containerd or crio) and Kubernetes cluster upgrade and timelines. So in that case, what is the expected behavior when:

1. containerd or crio version that includes this feature is running but Kuberenetes version running, does not include this feature
2. kubernetes version running includes this feature but containerd or crio version running, does not include this feature

Reply:
For this feature to work both containerd/crio and kubernetes would need to support the feature. Here's what should happen for both cases:
1. since kuberenetes does not include the feature you cannot create a Pod with the security context that has the ambient capabilities field. 
2. since containerd does not have this feature it should ignore the unknown field in the cri api object. Here is where we would update the documentation to call out that if you are enabling this feature on kubelet you would need containerd/crio version >= x.

Reply:
If I try to run a Pod that defines ambient capabilities and that Pod gets scheduled to a node that can't support ambient capabilities:
1. does the whole Pod fail or just the specific container that wanted ambient capabilities?
2. does an Event get created to explain that failure (if so, what event?)
3. does the kubelet keep trying to restart the failed Pod / container?

Reply:
Should we define a taint / toleration / [well known label](https://kubernetes.io/docs/reference/labels-annotations-taints/) to mark nodes that can support ambient capability requests?

Reply:
> If I try to run a Pod that defines ambient capabilities and that Pod gets scheduled to a node that can't support ambient capabilities:

> 1. does the whole Pod fail or just the specific container that wanted ambient capabilities?
> 2. does an Event get created to explain that failure (if so, what event?)
> 3. does the kubelet keep trying to restart the failed Pod / container?

Since, capabilities are defined per container the container would fail. kubelet would keep retrying the failed container. As this is in Alpha and is opt-in I think that is fine since we would document the supported runtime versions with the feature-flag. But when this feature is on-by default we would probably need to surface why the error hit in kubelet.

> Should we define a taint / toleration / well known label to mark nodes that can support ambient capability requests?

Excellent idea we should definitely do it in beta!